core/base/inc/ROOT/TExecutorCRTP.hxx

```diff
@@ -0,0 +1,352 @@
+   template< class F, class... T>
+   using noReferenceCond = typename std::enable_if<"Function can't return a reference" && !(std::is_reference<typename std::result_of<F(T...)>::type>::value)>::type;
+
+   // Map
+   // These trailing return types allow for a compile time check of compatibility between function signatures and args,
+   // and a compile time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F, class Cond = noReferenceCond<F>>
+   auto Map(F func, unsigned nTimes) -> std::vector<typename std::result_of<F()>::type>;
+   template<class F, class INTEGER, class Cond = noReferenceCond<F, INTEGER>>
+   auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector<typename std::result_of<F(INTEGER)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::initializer_list<T> args) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, const std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+
+   // MapReduce
+   // The trailing return types also check at compile time whether redfunc is compatible with func,
+   // other than checking that func is compatible with the type of arguments.
+   // a static_assert check in TExecutorCRTP<subc>::Reduce is used to check that redfunc is compatible with the type returned by func
+   template<class F, class R, class Cond = noReferenceCond<F>>
+   auto MapReduce(F func, unsigned nTimes, R redfunc) -> typename std::result_of<F()>::type;
+   template<class F, class INTEGER, class R, class Cond = noReferenceCond<F, INTEGER>>
+   auto MapReduce(F func, ROOT::TSeq<INTEGER> args, R redfunc) -> typename std::result_of<F(INTEGER)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, std::initializer_list<T> args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, const std::vector<T> &args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto MapReduce(F func, std::vector<T> &args, R redfunc) -> typename std::result_of<F(T)>::type;
+   template<class F, class T,class Cond = noReferenceCond<F, T>>
+   T* MapReduce(F func, std::vector<T*> &args);
+   template<class F, class T,class Cond = noReferenceCond<F, T>>
+   T* MapReduce(F func, const std::vector<T*> &args);
+
+   template<class T> T* Reduce(const std::vector<T*> &mergeObjs);
+   template<class T, class R> auto Reduce(const std::vector<T> &objs, R redfunc) -> decltype(redfunc(objs));
```

Comment:
You will have to ask the Enrico Guiraud from summer 2005 😅

Reply:
ugh

Reply:
Oh, because of the other reduce signature that doesn't take a reduction function, duh.

Maybe Map is the one who should have taken the map function as second argument 🤷‍♂️

Reply:
anyways...marking as "resolved" :D